presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java

```diff
@@ -333,6 +345,15 @@ public TaskInfo updateTask(Session session, Optional<PlanFragment> fragment, Lis
                 taskExecution = taskHolder.getTaskExecution();
                 if (taskExecution == null) {
                     checkState(fragment.isPresent(), "fragment must be present");
+                    // this block must happen before creating task execution to prevent a race condition when execution finished,
+                    // but types are not ready. Types are accessed by a task resource in order to convert pages to a client output
+                    if (fragment.get().getRoot() instanceof OutputNode) {
```

Comment:
My understanding is that this whole synchronized block happens only once per task because it creates task execution. However, I added a check as you suggested.